hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/util/Caches.java

```diff
@@ -291,6 +291,10 @@ public static void removeAll(AdvancedCache cache) {
 	}
 
 	public static CollectableCloseableIterable keys(AdvancedCache cache) {
+		if (cache.getCacheConfiguration().transaction().transactionMode().isTransactional()) {
+			// Dummy read to enlist the LocalTransaction as workaround for ISPN-5676
+			cache.containsKey(false);
```

Comment:
Wondering maybe you want to do a withFlags(CACHE_MODE_LOCAL) ?  So this way if it is distributed you won't incur the remote overhead.